api/src/main/java/org/apache/iceberg/TableScan.java

```diff
@@ -165,6 +166,22 @@ default TableScan select(String... columns) {
    */
   CloseableIterable<FileScanTask> planFiles();
 
+  /**
+   * Plan the {@link FileScanTask files} for the specified {@link ManifestFile manifests} that will be read by this
+   * scan.
+   * <p>
+   * Each file has a residual expression that should be applied to filter the file's rows.
+   * <p>
+   * This simple plan returns file scans for each file from position 0 to the file's length. For
+   * planning that will combine small files, split large files, and attempt to balance work, use
+   * {@link #planTasks()} instead.
+   *
+   * @param manifests the specified manifests for this scan.
+   * @param filterAddedFiles whether to filter just added files or all files.
+   * @return an Iterable of file tasks that are required by this scan
+   */
+  CloseableIterable<FileScanTask> planFilesForManifests(List<ManifestFile> manifests, boolean filterAddedFiles);
```

Comment:
This is something that a `ManifestGroup` should be used for instead. I don't think we want to add a new method to `TableScan`.

Reply:
How about removing the changes to scans and using the same underlying planning instead?

Here's a class that uses `ManifestGroup` just like the incremental scans. You can use it to get a `CloseableIterable` of `FileScanTask` for a single Snapshot, either full or just appends:

```java
package org.apache.iceberg;

import com.google.common.collect.Iterables;
import org.apache.iceberg.expressions.Expression;
import org.apache.iceberg.io.CloseableIterable;
import org.apache.iceberg.io.FileIO;

class Snapshots {
  public IterableBuilder from(Snapshot snapshot, FileIO io) {
    return new IterableBuilder(io, snapshot);
  }

  public class IterableBuilder {
    private final FileIO io;
    private final Snapshot snapshot;
    private Expression rowFilter;

    private IterableBuilder(FileIO io, Snapshot snapshot) {
      this.io = io;
      this.snapshot = snapshot;
    }

    public IterableBuilder filter(Expression newRowFilter) {
      this.rowFilter = newRowFilter;
      return this;
    }

    public CloseableIterable<FileScanTask> full() {
      return new ManifestGroup(io, snapshot.manifests())
          .filterData(rowFilter)
          .planFiles();
    }

    public CloseableIterable<FileScanTask> appends() {
      Iterable<ManifestFile> newManifests = Iterables.filter(snapshot.manifests(),
          manifest -> manifest.snapshotId() == snapshot.snapshotId());

      return new ManifestGroup(io, newManifests)
          .filterData(rowFilter)
          .filterManifestEntries(entry -> entry.snapshotId() == snapshot.snapshotId())
          .ignoreDeleted()
          .ignoreExisting()
          .planFiles();
    }
  }
}
```

The advantage here is that by calling `ManifestGroup` directly, you don't need to change how scans work at all -- so no need to change how the threading works, although that would be a great independent PR to open.

This gives you deterministic access to the files you need from a snapshot. For a starting snapshot, you'd get a `full()` task iterable and use the offset to filter. For other snapshots, you'd get the `appends()` task iterable and do the same offset filtering. I think this will make the implementation a lot more compact.

Reply:
Sorry for the late response @rdblue , I will take a look at this implementation today.